### PR TITLE
[factory] simplify interface for archive mode

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
make the Factory interface more natural to support archive mode
1. functions needed to support archive-mode are now implemented by the new `WorkingSetAtHeight()` interface
2. removed 2 duplicate methods and obsolete `historyfactory.go`

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/managers.go [hunk 1]
```diff
@@ -86,6 +89,14 @@ type (
 		Dock
 	}
 
+	// ArchiveStateSimulator defines an interface to read state and
+	// simulate-run actions on a given height
+	ArchiveStateSimulator interface {
+		StateReader
+		SimulateExecution(context.Context, address.Address, action.Envelope, ...SimulateOption) ([]byte, *action.Receipt, error)
+		ReadContractStorage(context.Context, address.Address, []byte) ([]byte, error)
```
Comment (dustinxie) [inline action/protocol/managers.go:97]: these 2 are needed for archive-node API

File: state/factory/factory.go [hunk 2]
```diff
@@ -485,52 +499,6 @@ func (sf *factory) DeleteTipBlock(_ context.Context, _ *block.Block) error {
...
-		return nil, err
-	}
-	defer tlt.Stop(context.Background())
-	keys, values, err := readStatesFromTLT(tlt, cfg.Namespace, cfg.Keys)
-	if err != nil {
-		return nil, err
-	}
-	return state.NewIterator(keys, values)
-}
-
```
Comment (dustinxie) [inline state/factory/factory.go:533]: move to `factoryWithHeight`

File: state/factory/factory_withheight.go
Comment (envestcc) [inline state/factory/factory_withheight.go:1]: `SimulateExecution` and `ReadContractStorage` also need re-implement, they are not currently based on the specified height, but rather on the latest height

File: state/factory/factory_withheight.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+type factoryWithHeight struct {
+	*factory
+	height uint64
+}
+
+func (sf *factoryWithHeight) State(s interface{}, opts ...protocol.StateOption) (uint64, error) {
+	sf.mutex.RLock()
+	defer sf.mutex.RUnlock()
```
[the remark this answers is not in the document]
Comment (envestcc) [reply state/factory/factory_withheight.go:25]: yeah, I think reading on archive state can be lock-free

File: state/factory/factory_withheight.go
Comment (dustinxie) [reply state/factory/factory_withheight.go:1]: ok just added, was thinking to add it later in API PR

File: state/factory/factory_withheight.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply state/factory/factory_withheight.go]: updated in latest commit

File: state/factory/factory.go [hunk 4]
```diff
@@ -438,9 +429,7 @@ func (sf *factory) ReadContractStorage(ctx context.Context, contract address.Add
 
 // PutBlock persists all changes in RunActions() into the DB
 func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) error {
-	sf.mutex.Lock()
 	timer := sf.timerFactory.NewTimer("Commit")
-	sf.mutex.Unlock()
```
Comment (dustinxie) [inline state/factory/factory.go:429]: when reviewing the usage of `sf.mutex`, found it's not really needed here?

File: state/factory/factory_withheight.go [hunk 5]
```diff
@@ -21,17 +25,18 @@ type factoryWithHeight struct {
...
 	if err != nil {
 		return 0, err
 	}
 	if cfg.Keys != nil {
 		return 0, errors.Wrap(ErrNotSupported, "Read state with keys option has not been implemented yet")
 	}
-	if sf.height > sf.currentChainHeight {
-		return 0, errors.Errorf("query height %d is higher than tip height %d", sf.height, sf.currentChainHeight)
+	sf.mutex.RLock()
+	tipHeight := sf.currentChainHeight
```
Comment (dustinxie) [inline state/factory/factory_withheight.go:36]: access to `sf.currentChainHeight` should be protected by `sf.mutex`

File: state/factory/factory_withheight.go [hunk 6]
```diff
@@ -21,25 +25,17 @@ type factoryWithHeight struct {
...
 	cfg, err := processOptions(opts...)
 	if err != nil {
 		return 0, err
 	}
 	if cfg.Keys != nil {
 		return 0, errors.Wrap(ErrNotSupported, "Read state with keys option has not been implemented yet")
 	}
-	if sf.height > sf.currentChainHeight {
-		return 0, errors.Errorf("query height %d is higher than tip height %d", sf.height, sf.currentChainHeight)
-	}
```
Comment (dustinxie) [inline state/factory/factory_withheight.go]: move the check to `AtHeight()`

File: state/factory/factory_withheight.go [hunk 7]
```diff
@@ -21,25 +25,17 @@ type factoryWithHeight struct {
...
 	return sf.height, sf.stateAtHeight(sf.height, cfg.Namespace, cfg.Key, s)
 }
 
 func (sf *factoryWithHeight) stateAtHeight(height uint64, ns string, key []byte, s interface{}) error {
-	if !sf.saveHistory {
-		return ErrNoArchiveData
-	}
```
Comment (dustinxie) [inline state/factory/factory_withheight.go]: move the check to `AtHeight()`

File: state/factory/factory_withheight.go [hunk 8]
```diff
@@ -57,11 +53,6 @@ func (sf *factoryWithHeight) stateAtHeight(height uint64, ns string, key []byte,
 }
 
 func (sf *factoryWithHeight) States(opts ...protocol.StateOption) (uint64, state.Iterator, error) {
-	sf.mutex.RLock()
-	defer sf.mutex.RUnlock()
-	if sf.height > sf.currentChainHeight {
-		return 0, nil, errors.Errorf("query height %d is higher than tip height %d", sf.height, sf.currentChainHeight)
-	}
```
Comment (dustinxie) [inline state/factory/factory_withheight.go]: move the check to `AtHeight()`

File: api/coreservice.go [hunk 9]
```diff
@@ -961,7 +961,11 @@ func (core *coreService) readState(ctx context.Context, p protocol.Protocol, hei
 		}
 		if inputHeight < tipHeight {
 			// old data, wrap to history state reader
-			d, h, err := p.ReadState(ctx, factory.NewHistoryStateReader(core.sf, inputHeight), methodName, arguments...)
+			historySR, err := core.sf.WorkingSetAtHeight(ctx, inputHeight)
+			if err != nil {
+				return nil, 0, err
+			}
+			d, h, err := p.ReadState(ctx, historySR, methodName, arguments...)
```
Comment (envestcc) [inline api/coreservice.go:968]: can combine codes after getting the specified StateReader

File: state/factory/factory.go [hunk 10]
```diff
@@ -268,7 +266,16 @@ func (sf *factory) newWorkingSet(ctx context.Context, height uint64) (*workingSe
...
+	var (
+		rootKey    = ArchiveTrieRootKey
+		createTrie = true
+	)
+	if height < sf.currentChainHeight {
+		// archive mode
+		rootKey = fmt.Sprintf("%s-%d", ArchiveTrieRootKey, height)
+		createTrie = false
+	}
+	store, err := newFactoryWorkingSetStore(sf.protocolView, flusher, rootKey, createTrie)
```
Comment (envestcc) [inline state/factory/factory.go:278]: Splitting into the following two functions seems more clear:
- `newFactoryWorkingSetStore(view, flusher)`, which inner use ArchiveTrieRootKey and true
- `newFactoryWorkingSetStoreAt(view, flusher, height)`, which inner use ArchiveTrieRootKey-height and false

Comment (envestcc) [reply state/factory/factory.go:278]: does it should use ArchiveTrieRootKey or ArchiveTrieRootKey-height if height equals currentChainHeight?

File: state/factory/workingset.go [hunk 11]
```diff
@@ -333,7 +333,7 @@ func (ws *workingSet) Commit(ctx context.Context) error {
 		return err
 	}
 	ws.Reset()
-	return nil
+	return ws.store.Stop(ctx)
```
Comment (envestcc) [inline state/factory/workingset.go:336]: why add this?

File: state/factory/workingset.go [hunk 12]
```diff
@@ -343,6 +343,9 @@ func (ws *workingSet) State(s interface{}, opts ...protocol.StateOption) (uint64
 	if err != nil {
 		return ws.height, err
 	}
+	if cfg.Keys != nil {
```
Comment (envestcc) [inline state/factory/workingset.go:346]: why add this?

File: state/factory/factory.go (hunk 10 above)
Comment (dustinxie) [reply state/factory/factory.go:278]: 2 cases for height equals currentChainHeight
1. createGenesisStates calls `newWorkingSet(ctx, 0)`, which is write at current tip
2. archive-mode trying to read at current tip height > 0, at the same time `PutBlock` may still be in progress so `ArchiveTrieRootKey-height` may not exist at the moment
both cases should use `ArchiveTrieRootKey `

File: state/factory/workingset.go (hunk 11 above)
Comment (dustinxie) [reply state/factory/workingset.go:336]: newWorkingSet() creates a `factoryWorkingSetStore` instance and calls `Start()` inside, but its `Stop()` is never called.

Comment (dustinxie) [reply state/factory/workingset.go:336]: After checking, `workingSet.Commit()` calls `store.tlt.RootHash()`, which calls `tlt.flush()`, which is the effective Stop(), we may need another PR to clean-up this

File: state/factory/workingset.go (hunk 12 above)
Comment (dustinxie) [reply state/factory/workingset.go:346]: it's in the removed `stateAtHeight()` func, it's a safety check, there's a similar check in `States()` at L

**Review by envestcc — APPROVED** [review]

**Review by huangzhiran — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**